homeassistant/components/elkm1/alarm_control_panel.py

```diff
@@ -61,93 +64,61 @@
     }
 )
 
+_LOGGER = logging.getLogger(__name__)
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the ElkM1 alarm platform."""
-    if discovery_info is None:
-        return
 
-    elk_datas = hass.data[DOMAIN]
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the ElkM1 alarm platform."""
+    elk_data = hass.data[DOMAIN][config_entry.entry_id]
     entities = []
-    for elk_data in elk_datas.values():
-        elk = elk_data["elk"]
-        entities = create_elk_entities(elk_data, elk.areas, "area", ElkArea, entities)
-    async_add_entities(entities, True)
 
-    def _dispatch(signal, entity_ids, *args):
-        for entity_id in entity_ids:
-            async_dispatcher_send(hass, f"{signal}_{entity_id}", *args)
+    elk = elk_data["elk"]
+    areas_with_keypad = set()
+    for keypad in elk.keypads:
+        areas_with_keypad.add(keypad.area)
 
-    def _arm_service(service):
-        entity_ids = service.data.get(ATTR_ENTITY_ID, [])
-        arm_level = _arm_services().get(service.service)
-        args = (arm_level, service.data.get(ATTR_CODE))
-        _dispatch(SIGNAL_ARM_ENTITY, entity_ids, *args)
+    areas = []
+    for area in elk.areas:
+        if area.index in areas_with_keypad or elk_data["auto_configure"] is False:
+            areas.append(area)
+    create_elk_entities(elk_data, areas, "area", ElkArea, entities)
+    async_add_entities(entities, True)
 
-    for service in _arm_services():
-        hass.services.async_register(
-            DOMAIN, service, _arm_service, ELK_ALARM_SERVICE_SCHEMA
-        )
+    platform = entity_platform.current_platform.get()
 
-    def _display_message_service(service):
-        entity_ids = service.data.get(ATTR_ENTITY_ID, [])
-        data = service.data
-        args = (
-            data["clear"],
-            data["beep"],
-            data["timeout"],
-            data["line1"],
-            data["line2"],
+    for service in ARM_SERVICES:
```

Comment:
This saves 0 lines of code versus writing it out 😆

Reply:
I'll unroll the loop.